### PR TITLE
Add minimum coil tier required to JEI information.

### LINKED ISSUE #1342: [Suggestion] Coil tier indication in JEI
**Is your feature request related to a problem? Please describe.**
For Blast Furnace recipes, it would be nice to quickly see what coil tier (at the minimum) is needed

**Describe the solution you'd like**
Similar to the power tier indicator, add the coil tier after the temperature, e.g. **Temperature: 2900K (Nichrome)**

### PR DESCRIPTION
**Pull request by detav**
**What:**
Adds the name of the minimum required coil to JEI as suggested by #1342

**Outcome:**
Closes #1342.

**Additional info:**
Preview:

Some tiers break the tab border, might be inconvenient:

### CODE REVIEW

**Review by Archengius — CHANGES_REQUESTED** [review]
Getting blast furnace specific code into the general JEI recipe handling routine is a bad idea because it only solves one particular issue and doesn't allow any kind of extensibility beyond that.

As LAGIdiot said earlier in the original issue, it would be a better idea to overhaul system in question entirely, but I guess it's more of a long term task.

My current suggestion would be adding client side only method in RecipeMap that would be responsible for adding property-related tooltips to recipes, and override it for blast furnace recipe map.

**Comment by Archengius** [comment]
As of the long-term solution, I would suggest replacing string keys in property map with a dedicated generic base class that will define type of the property value strictly and handle it's formatting/etc.

Now I'm thinking about it, and it doesn't sound like a complex thing to implement. Pretty sure you can get it working even in scope of this PR. Backwards compatibility might be an issue though, so I would leave string-related methods in case and remap them to new system by names in runtime.
Now im

**Review by LAGIdiot — CHANGES_REQUESTED** [review]
I completely agree with Arch. You should not use that specific logic in general place like JEI.

So please rework this PR either with method in RecipeMap (easy way) or if you feel like it you can go with suggested long-term solution.

**Review by LAGIdiot — CHANGES_REQUESTED** [review]
Good work on this I have added some comments regarding how I would like to see this now that I have more concreate idea about how this can be done.

**Review by Archengius — CHANGES_REQUESTED** [review]
You got my initial idea wrong. I was talking about making properties strongly typed by making keys carry more information than just a name. They would define value class (ideally by using generics) and handle rendering of the provided value. But they don't define value itself. Their purpose is to be used as a key to look up value inside of the recipe while providing additional handling of it and ensuring type safety.

**Comment by LAGIdiot** [comment]
Are you still working on this? Need some help with it?

File: src/main/java/gregtech/api/recipes/Recipe.java
Comment (detav) [reply src/main/java/gregtech/api/recipes/Recipe.java:74]: Right now `TemperatureProperty` is in the common package because `CoilType` is in there, for this I would have to move it to api, and I am not sure how to do that. What do you suggest?

Comment (LAGIdiot) [reply src/main/java/gregtech/api/recipes/Recipe.java:74]: I don't see any reasons why `TemperatureProperty` should be in API. If this function takes `RecipeProperties` then as long `TemperatureProperty` extends it then it can be passed in.

**Comment by LAGIdiot** [comment]
I have been looking to this for more then 2 hours and I am completely lost. I don't remember how I was imagining to get this implemented and from comments I and Arch left I can't piece it together. I feel like we mixed two different concepts together and based on them you implemented it.

I will have to take another look at this later. For now I will open this for discussion with hope that someone will be able to provide insight or idea which will help me to get some solid hold on this.

**Comment by detav** [comment]
Maybe I should have updated the description.
I decided to go for Archengius implementation, just that. The rest is the actual TemperatureProperty which based on registered coil types renders the minimum coil type needed for the recipe's property value (the temperature).

**Comment by ALongStringOfNumbers** [comment]
As an explanation for @LAGIdiot There are several parts to this PR.

## Part 1: Creating the new Property
Building off of the comment from Arch here:
> As of the long-term solution, I would suggest replacing string keys in property map with a dedicated generic base class that will define type of the property value strictly and handle it's formatting/etc.

The new Property class is defined with a combination of `BaseProperty` and `RecipeProperty`. `BaseProperty` is used to remap existing properties into the new system and provide a default implementation for anything not changed in this PR. As this is meant for updating old properties, it takes a String key specifically, as that is what is currently implemented. Note that GTCE itself doesn't really need this, as its one multiblock with a property that would be useful for displaying, the EBF, is already updated in this PR, however it will be useful for addons if they do not update immediately after the PR is merged and released.

  `RecipeProperty` on the other hand is the generic version of and new implementation of `BaseProperty` as mentioned in the comment by Arch above. Instead of solely handling `String` various types of `RecipeProperty` can be defined, see `TemperatureProperty` for an example with an integer. The `key` in `RecipeProperty` is used as a unique identifier, so that each property can be differentiated from each other, but still be generic and not have to be passed in a `String` in the declaration of the property

## Part 2: Remapping existing Properties
With the switch to generic properties, the existing hardcoded `String` properties need to be remapped into the new system. This is mainly handled in `Recipe`. The Constructor of `Recipe` is changed to take in a map of a `generic` and `object` instead of the previously hardcoded `String` and `Object`. This is most likely done to preserve compatibility with addons, however an additional option is to create a secondary constructor that takes a `Map<RecipeProperty<?>, Object>` instead of a `Map<String, Object>` to ensure that `Recipe` will be passed a `RecipeProperty` and change the existing constructor to call the newer constructor after remapping the passed `Map<String, Object>`, similar to how I briefly implemented it in #1502.

Either option would see the existing properties passed into `Recipe` be remapped into the new system using the new `remapProperties` method and then stored inside an instance variable, or not remapped and just stored if it is already in the new implementation.

The other changes in `Recipe` are simply switching over the existing getters for the `RecipeProperty` so that they no longer are based on the idea of a hardcoded `String` property. This is also where the unique `key` defined for each property comes into play, as it allows for easy identification of the property based on the key.

## Part 3: Implementing new Properties for the EBF
The final part of this PR is implementing this new system for the Electric Blast Furnace, which I believe to be the only structure in GTCE that needs the new implementation, the rest of the structures that could use the implementation are in addon mods.

A new `TemperatureProperty` is created (Note this has to be in API since the `BlastRecipeBuilder` is also in API). `TemperatureProperty` gets a list of coils when recipes are being registered (As CoilType is in Common, so it cannot be called directly). The coils are then stored along with the heat value they handle and their material. The two methods from `RecipeProperty` are overridden for the new implementation, namely the unique identifier and displaying the correct information on the JEI page, where the information on the JEI page finds the minimum coil for the provided temperature in the recipe.

In addition to this new `TemperatureProperty`, `GTRecipeWrapper` is changed to use this new implementation instead of being hardcoded to strings.

I hope that helps explain some of how this PR works, it is not mine so maybe @detav has some more information/notes on how this PR works.


Some questions for @detav now though.
1. Was this tested for the property of `<String, ItemStack>` that was added to the Implosion Compressor a while back? I imagine there would be no issues due to how the remapping is done, but I have not checked.
2. How would an addon Mod, such as Gregicality, add their additional coil types to the list of coil types to be displayed in JEI? Would they be able to use the existing registration method in `TemperatureProperty`, filtering out the coils that currently exist in GTCE, or would it be beneficial to add a method to clear the stored coils, so that addons can register everything together, which might also be useful if an addon mod changes the temperature that a specific coil can handle.

**Comment by detav** [comment]
Amazing, very comprehensive, thanks.

> however an additional option is to create a secondary constructor that takes a Map<RecipeProperty<?>, Object> instead of a Map<String, Object>

I can't do that, the java compiler treats them as identical so I had to change the current constructor.

> the Electric Blast Furnace, which I believe to be the only structure in GTCE that needs the new implementation

There is also the [Fusion Reactor](https://github.com/GregTechCE/GregTech/blob/f0ee6248a7ee29177b326830561ffd1176ca7e75/src/main/java/gregtech/api/recipes/builders/FusionRecipeBuilder.java#L54), for example, which I don't know if addons use but if something wants to be added would have to be in another pr I guess.

> Was this tested for the property of <String, ItemStack> that was added to the Implosion Compressor a while back?

The Implosion Compressor [doesn't pass any property to the recipe](https://github.com/GregTechCE/GregTech/blob/f0ee6248a7ee29177b326830561ffd1176ca7e75/src/main/java/gregtech/api/recipes/builders/ImplosionRecipeBuilder.java#L87).

> How would an addon Mod, such as Gregicality, add their additional coil types to the list of coil types to be displayed in JEI?

The current  method uses a TreeMap so newly added coils should be ordered fine. Not sure if they can change the temperatures as the coils are enums? They would have to change the temperatures before the coils are registered.
I will test this with gregicality.

**Comment by ALongStringOfNumbers** [comment]
>There is also the Fusion Reactor, for example, which I don't know if addons use but if something wants to be added would have to be in another pr I guess.

As GTCE does not implement anything for the Fusion Reactor besides creating the RecipeBuilder, I would say it is fine to leave it to addons to implement the RecipeProperty for the Fusion Reactor.

**Comment by LAGIdiot** [comment]
@ALongStringOfNumbers thanks for explanation it helped.

As we already mentioned we were thinking about having `RecipeProperty` as blueprint for formatting data but not keeping them. Which would make sense to have known types as Singleton as there is no need for more instances. On other hand we need to have system which will convert incoming data from old way `string-value` to new system and for these we need new instance every time as you we need to save `property name` in it. But that would mean possible many instances with same `property name` which we are trying to mitigate... But that should go away as addons switch to new system.
It would be better to leave old constructor as deprecated and handle conversion. And new one which will take `Map of RecipeProperty-value` so it is explicit that there is some change. It may need to have check that passed `RecipeProperty` can work with provided value.
Way type inside is accessed is way too open it would be better to have it private and access it though function like `isOfType` which would take type in and return `boolean`. Same goes for casting that should be handled by `RecipeProperty` too with proper error handling.
All our calls to `getIntegerProperty()` (and similar functions) should be migrated to new `getPropertyValue()` and marked as deprecated.
`RecipeProperty.drawInfo()` should not get whole recipe but just appropriate value (as object and cast it it self) so there is no additional coupling. Which means that `GTRecipeWrapper.drawInfo()` will need to work on top of different set.

**Comment by detav** [comment]
I addressed the requests that I could. I also changed the names of the properties, should be better now.

> It would be better to leave old constructor as deprecated and handle conversion. And new one which will take Map of RecipeProperty-value so it is explicit that there is some change. It may need to have check that passed RecipeProperty can work with provided value.

As I said, I can't make 2 constructors where the only difference is the parameter types of the map, it won't compile.

**Comment by LAGIdiot** [comment]
I have been going through this and trying some things and I have prototype of even better implementation then I proposed. I will clean it up a bit, commit it and then we can build on top of that.

**Comment by LAGIdiot** [comment]
This proved to be more complicated then I imagine because of compatibility and feature set. I just need to covert some of our Builders to new system and that should be it from code perspective.

**Comment by LAGIdiot** [comment]
Thanks for your hard work
Superseded by #1580

**Pull request closed without merging**
